### PR TITLE
Recipe to add a new language to Llama2

### PR DESCRIPTION
**Pull request by rahul-sarvam**
# What does this PR do?

This PR introduces a new recipe to augment a new language to the Llama2 tokenizer and continue pre-training. This recipe replicates the results of [OpenHathi](https://www.sarvam.ai/blog/announcing-openhathi-series) when the right data mixture is used.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: recipes/multilingual/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# Extending Llama to a new language
+
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:2]:
```suggestion
Authored by : Saravam team
```

Comment (HamidShojanazeri) [reply recipes/multilingual/README.md:2]: or your name any one that fits better

File: recipes/multilingual/README.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+In this recipe, we will see how to add a new language to the Llama family of models. The steps are quite general and can be easily adapted to other models as well. Using this recipe, you should be able to replicate the findings of [OpenHathi](https://huggingface.co/sarvamai/OpenHathi-7B-Hi-v0.1-Base).
+
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:4]:
```suggestion
Please read more about OpenHathi [here](https://www.sarvam.ai/blog/announcing-openhathi-series)
```

File: recipes/multilingual/README.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+>>> text = "मैं एक अच्छा हाथी हूँ"
+>>> llama_tokenizer.tokenize(text)
+['▁', 'म', 'ै', 'ं', '▁', '<0xE0>', '<0xA4>', '<0x8F>', 'क', '▁', 'अ', 'च', '्', '<0xE0>', '<0xA4>', '<0x9B>', 'ा', '▁', 'ह', 'ा', 'थ', 'ी', '▁', 'ह', 'ू', '<0xE0>', '<0xA4>', '<0x81>']
+>>> our_tokenizer.tokenize(text)
+['▁मैं', '▁एक', '▁अच', '्', 'छा', '▁हाथी', '▁हूँ']
+```
+
+## Continual pre-training
+OpenHathi uses a two-stage pre-training process:
+- Phase 1: learn to translate paragraphs of text (use translated text as context and generate the original text)
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:57]: can we please add how much data we would need/ produce in phase 1 and 2?

File: recipes/multilingual/README.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+- Phase 2: bilingual next token prediction (train on text where the language changes after every sentence)
+
+We can easily create data for both phases using any translation model. OpenHathi uses [IndicTrans2](https://github.com/AI4Bharat/IndicTrans2). We provide sample code for both phases below.
+
+### Phase 1
+With the assumption that we don't have source-native data, let us first get some English data to translate. 
+
+```
+from datasets import load_dataset
+ds = load_dataset("rahular/varta", split="validation", streaming=True)
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:67]: wonder if this would be validation set or train? in terms of supporting enough data for training.

Can we please also have some hint how to save this data in what format to use later, would it be HF dataset etc.?

File: recipes/multilingual/README.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+ip = IndicProcessor(inference=True)
+
+phase1_data = []
+for para in english_paragraphs:
+    trans_para = translate_paragraph(para, "eng_Latn", "hin_Deva", en_indic_model, en_indic_tokenizer, ip)
+    phase1_data.append({"text": f"{trans_para}\n\n{para}"})
+```
+
+### Phase 2
+This is almost the same as phase 1, except that we have to replace the original sentences in an alternating manner to get the data in the required format. We can use the `split_sentences` ([link](https://github.com/AI4Bharat/IndicTrans2/blob/main/huggingface_interface/example.py#L60])) and `batch_translate` ([link](https://github.com/AI4Bharat/IndicTrans2/blob/main/huggingface_interface/example.py#L109)) functions to do this.
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:89]: similarly here, can we also have hint/ add code how to save the dataset/ format.

File: recipes/multilingual/README.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    trans_sents = batch_translate(input_sentences, "eng_Latn", "hin_Deva, en_indic_model, en_indic_tokenizer, ip)
+    final_para = []
+    for idx, (en_sent, trans_sent) in enumerate(zip(en_sents, trans_sents)):
+        sent_to_append = en_sent if idx % 2 == 0 else trans_sent
+        final_para.append(sent_to_append)
+    phase2_data.append({"text": " ".join(final_para)})
+```
+
+### Train
+Finally, we can start finetuning Llama2 on these datasets by following the [finetuning recipes](https://github.com/rahul-sarvam/llama-recipes/tree/main/recipes/finetuning). Remember to pass the new tokenizer path as an argument to the script: `--tokenizer_name=./extended_tokenizer`.
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:109]:
```suggestion
Finally, we can start finetuning Llama2 on these datasets by following the [finetuning recipes](https://github.com/meta-llama/llama-recipes/tree/main/recipes/finetuning). Remember to pass the new tokenizer path as an argument to the script: `--tokenizer_name=./extended_tokenizer`.
```

**Review by HamidShojanazeri — COMMENTED** [review]
Thanks a lot @rahul-sarvam , it looks great! just added some minor comments.

File: recipes/multilingual/README.md (hunk 3 above)
Comment (rahul-sarvam) [reply recipes/multilingual/README.md]: Added in 09028bf

File: recipes/multilingual/README.md (hunk 4 above)
Comment (rahul-sarvam) [reply recipes/multilingual/README.md]: Changed in 09028bf

File: recipes/multilingual/README.md (hunk 5 above)
Comment (rahul-sarvam) [reply recipes/multilingual/README.md:96]: Added in 09028bf

**Review by HamidShojanazeri — APPROVED** [review]
